### DISCUSSION TITLE
Bootstrapping (Technical Questions)

### DISCUSSION THREAD

**Discussion by cyhsieh-psy** [RePsychLing/SMLP2023#31] [Q&A]
1. Are there ways to save all the output (both figures and numbers)? 

2. Are there parallel processing in julia to speed up the boostrapping procedure?

**Comment by palday** [comment]
1. Figures can be saved with the usual saving mechanisms in Julia/Makie -- look at the documentation for `save`. For the bootstrap replicates, there is a function [`savereplicates`](https://juliastats.org/MixedModels.jl/v4.22/api/#MixedModels.savereplicates-Tuple{Any,%20MixedModels.MixedModelFitCollection}).
2. Yes, but it's not trivial to use and often not worthwhile for a typical laptop / desktop. Earlier versions of `parametricbootstrap` had an argument for running multiple bootstrap replicates in parallel with multi-threading. At first, this give a big performance boost, but as the linear algebra libraries (the bits of code that does all the matrix math) became better, this performance boost went away. Why? Because the linear algebra libraries are themselves multi-threaded and so the different levels of threads wound up competing with each other. This contention between threads negated any boost, so we decided to simplify the code and drop support for multi-threading. That said, we do provide a few utilities for combining separate bootstrap objects. This is useful if you have access to a cluster, in which case you may have multiple bootstraps running on multiple cluster nodes simultaneously. The `Distributed` standard library in Julia can help you with this, but the full details of setting this up on a cluster go beyond the scope of this core. The [documentation](https://juliastats.org/MixedModels.jl/v4.22/bootstrap/) includes [a minimal worked example of computing a bootstrap sample with `Distributed`](https://juliastats.org/MixedModels.jl/v4.22/bootstrap/#Distributed-Computing-and-the-Bootstrap).
